File: sdk/python/packages/flet/src/flet/controls/material/sliver_scroll_view.py
from typing import List

from flet.controls.adaptive_control import AdaptiveControl
from flet.controls.base_control import control
from flet.controls.material.sliver import Sliver
from flet.controls.types import Number, ClipBehavior, OptionalNumber, OptionalInt

__all__ = ["SliverScrollView"]


@control("SliverScrollView")
class SliverScrollView(AdaptiveControl, Sliver):
    """
    A scroll view that creates custom scroll effects using slivers.

    -----

    Online docs: https://flet.dev/docs/controls/sliverscrollview
    """

    slivers: List[Sliver]
    clip_behavior: ClipBehavior = ClipBehavior.HARD_EDGE
    reverse: bool = False
    horizontal: bool = False
    shrink_wrap: bool = False
    semantic_child_count: OptionalInt = None
    anchor: Number = 0.0
    primary: bool = False
    cache_extent: OptionalNumber = None

File: packages/flet/lib/src/controls/sliver_scroll_view.dart
import 'package:flet/src/extensions/control.dart';
import 'package:flet/src/utils/numbers.dart';
import 'package:flutter/material.dart';

import '../models/control.dart';
import '../utils/misc.dart';
import 'base_controls.dart';
import 'scroll_notification_control.dart';
import 'scrollable_control.dart';

class SliverScrollViewControl extends StatefulWidget {
  final Control control;

  const SliverScrollViewControl({super.key, required this.control});

  @override
  State<SliverScrollViewControl> createState() =>
      _SliverScrollViewControlState();
}

class _SliverScrollViewControlState extends State<SliverScrollViewControl> {
  late final ScrollController _controller = ScrollController();

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("SliverScrollView build: ${widget.control.id}");

    var ctrls = widget.control.buildWidgets("controls");
    var scrollDirection = widget.control.getBool("horizontal", false)!
        ? Axis.horizontal
        : Axis.vertical;

    Widget scrollView = CustomScrollView(
      clipBehavior:
          widget.control.getClipBehavior("clip_behavior", Clip.hardEdge)!,
      reverse: widget.control.getBool("reverse", false)!,
      scrollDirection: scrollDirection,
      shrinkWrap: widget.control.getBool("shrink_wrap", false)!,
      semanticChildCount:
          widget.control.getInt("semantic_child_count", ctrls.length),
      anchor: widget.control.getDouble("anchor", 0.0)!,
      primary: widget.control.getBool("primary"),
      cacheExtent: widget.control.getDouble("cache_extent"),
      slivers: ctrls,
    );
    Widget child = ScrollableControl(
        control: widget.control,
        scrollDirection: scrollDirection,
        scrollController: _controller,
        child: scrollView);

    if (widget.control.getBool("on_scroll", false)!) {
      child = ScrollNotificationControl(control: widget.control, child: child);
    }

    return BaseControl(control: widget.control, child: child);
  }
}
